[Enhancement] Refactor winter:env, winter:install to use new EnvFile & ArrayFile parsers

LukeTowers opened the pull request:
Related to wintercms/storm#40

TODO:
- [x] Refactor mail config mirroring to work like the database one now since the default mail config structure has changed.

<a href="https://gitpod.io/#https://github.com/wintercms/winter/pull/404"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

jaxwilko committed:
Updated winter install command to use ConfigFile

jaxwilko committed:
Re-implemented winter:env using EnvFile

jaxwilko committed:
Merge branch 'develop' into wip/config-file

jaxwilko committed:
Added fix to when loading config values to write in env file

jaxwilko committed:
Updated test case with all strings being quoted

LukeTowers commented:
@jaxwilko are these tests only failing because it's not pulling in the correct version of Storm?

jaxwilko commented:
@LukeTowers should be yes, storm can get merged first as it doesn't overwrite anything in use, then this branch should run the tests fine :)

LukeTowers committed:
Run default config files through ConfigWriter to minimize changed lines on future uses

LukeTowers committed:
Merge branch 'wip/1.2' into wip/config-file

# Conflicts:
#	modules/system/console/WinterEnv.php

LukeTowers committed:
Bring in line with changes to the storm PR

LukeTowers committed:
Fix botched merge

LukeTowers committed:
Fix winter:env test

LukeTowers committed:
Run the default config through ArrayFile parser

This serves the dual purpose of both testing the output of the ArrayFile parser and normalizing the configs so that any future uses of ArrayFile on them will have easier to understand diffs.

These were generated by running the following in Tinker:

use Winter\Storm\Parse\PHP\ArrayFile;

$configPaths = glob(base_path('config/*'));

foreach ($configPaths as $path) {
  if (is_file($path)) {
  	ArrayFile::open($path)->write();
  }
}

LukeTowers commented on config/app.php:
```diff
@@ -1,7 +1,7 @@
 
 return [
-
+    
```
@jaxwilko This whitespace should be removed by our printer

LukeTowers commented on config/app.php:
```diff
@@ -244,14 +243,12 @@
-    'providers' => array_merge(include(base_path('modules/system/providers.php')), [
-
+    
+    'providers' => array_merge(include base_path('modules/system/providers.php'), [
```
@jaxwilko looks like the printer / parser barfs on nested functions, we'll need to support that

jaxwilko replied:
Added with backtracing to ensure includes were () correctly dependent upon position: https://github.com/wintercms/storm/pull/40/commits/c098853f8e082e4a29e73c8dc21d873195337777
Removed above and simply always appended (): https://github.com/wintercms/storm/pull/40/commits/892ed2a1dd87e4bb75adce332fe96288fcf16260

LukeTowers commented on config/app.php:
```diff
@@ -244,14 +243,12 @@
         // 'Illuminate\Html\HtmlServiceProvider', // Example
-
-        'System\ServiceProvider',
+        'System\\ServiceProvider',
```
I'm not sure why this was changed, is it a fluke that the original value worked? Or is the printer misunderstanding that backslashes are only required to be escaped in single quote strings if they're right next to a value that would otherwise have to be escaped (i.e. a single quote)?

jaxwilko replied:
When you pass a variable it'll pass the value of that variable, i.e. if you were to pass `CURLOPT_CRLF` it would set the value 27.

To set a namespace you'd either need to use `\\\` as a separator or you could use the `constant()` method and set the full class reference, i.e. `Path\Name\MyClass::class`

LukeTowers commented on config/auth.php:
```diff
@@ -1,39 +1,41 @@
+                |
+                | If set to true, users will be given a limited number of attempts to sign
+                | in to the Backend before being blocked for a specified number of minutes.
+                |
```
@jaxwilko indentation is borked here, not sure why.

LukeTowers commented on config/logging.php:
```diff
@@ -42,74 +42,66 @@
-            'driver'       => 'monolog',
-            'handler'      => env('LOG_PAPERTRAIL_HANDLER', \Monolog\Handler\SyslogUdpHandler::class),
+            'driver' => 'monolog',
+            'handler' => env('LOG_PAPERTRAIL_HANDLER', \Monolog\Handler\SyslogUdpHandler::class),
```
@jaxwilko Laravel actually imports these classes at the top of their files (see https://github.com/laravel/laravel/blob/9.x/config/app.php#L3), which might be a setup that we could encounter (thus failing the file must start with return requirement), is it easy to support that or should we just say that we don't support it?

jaxwilko replied:
Should be fairly easy to check for and ignore imports prior to a return stmt

jaxwilko replied:
Support added in https://github.com/wintercms/storm/pull/40/commits/ec040435c486b0a5cb369137cf3d5dfa2d932c2e

LukeTowers commented on config/mail.php:
```diff
@@ -32,53 +32,46 @@
+            'port' => env('MAIL_PORT', 587),
+            'timeout' => null,
+            'transport' => 'smtp',
+            'username' => env('MAIL_USERNAME'),
```
@jaxwilko just to confirm, is trailing commas enforced in the ArrayPrinter?

jaxwilko replied:
It was at some point yes, if you check the early posts on the PR there was an example of it, it'd be worth looking at the tests to confirm though

jaxwilko replied:
Yes, it is

LukeTowers commented on config/session.php:
```diff
@@ -123,12 +125,9 @@
-    ),
-
+    
+    'cookie' => env('SESSION_COOKIE', str_slug(env('APP_NAME', 'winter'), '_') . '_session'),
```
Looks like this one came across properly, although it would be nice if it didn't eat the formatting. Easy fix or no?

jaxwilko replied:
You'd have to embed custom formatting into the printer, which, ehh, we can have a look but no promises 😂

LukeTowers committed:
Revert changes to winter:install, ConfigWriter uses ArrayFile internally

LukeTowers commented:
Blocked by comments, especially the issue with the `include()` broken parsing / rendering.

LukeTowers committed:
Revert "Run default config files through ConfigWriter to minimize changed lines on future uses"

This reverts commit 897263e2813596e116ed0b1350e1a9034a646cb0.

LukeTowers committed:
Merge branch 'wip/1.2' into wip/config-write-replacement

jaxwilko committed:
Merge branch 'wip/1.2' into wip/config-write-replacement

jaxwilko committed:
Merge branch 'wip/config-file' into wip/config-write-replacement

LukeTowers committed:
Merge branch 'wip/config-write-replacement' of github.com:wintercms/winter into wip/config-write-replacement

LukeTowers committed:
Revert "Run the default config through ArrayFile parser"

This reverts commit c3bdfae664c1cdde2ea1403027fc38b0822a0cf5.

LukeTowers committed:
Merge branch 'wip/1.2' into wip/config-write-replacement

LukeTowers committed:
Tweaks to be more friendly with the ArrayFile parser

LukeTowers committed:
Merge branch 'wip/config-write-replacement' into wip/config-file

# Conflicts:
#	config/app.php
#	config/auth.php
#	config/cms.php
#	config/database.php

LukeTowers committed:
Style fix

LukeTowers commented:
Blocked by remaining TODO item and https://github.com/wintercms/storm/pull/40#issuecomment-1068693492

LukeTowers committed:
Run all config files through ArrayFile parser

This will minimize changes when ConfigWriter is used to set config values through PHP.

LukeTowers committed:
Support new mail config file structure
